src/notebookConcatDocument.ts

```diff
@@ -1,302 +1,356 @@
+    this._selector = selector;
+    // Create a safe notebook document so that we can handle both >= 1.56 vscode API and < 1.56
+    // when vscode stable is 1.56 and both Python release and insiders can update to that engine version we
+    // can remove this and just use NotebookDocument directly
+    this._notebook = notebook;
+    // Note: Has to be different than the prefix for old notebook editor (HiddenFileFormat) so
+    // that the caller doesn't remove diagnostics for this document.
+    this.dummyFilePath = path.join(
+      dir,
+      `${NotebookConcatPrefix}${uuid().replace(/-/g, "")}.py`
+    );
+    this.dummyUri = Uri.file(this.dummyFilePath);
+
+    if (notebook.uri.scheme === InteractiveScheme) {
+      this.concatDocument = new InteractiveConcatTextDocument(
+        notebook,
+        selector,
+        notebookApi
+      );
+    } else {
+      this.concatDocument = new EnhancedNotebookConcatTextDocument(
+        notebook,
+        selector,
+        notebookApi
+      );
     }
 
-    public get fileName(): string {
-        return this.dummyFilePath;
+    this.onDidChangeSubscription = this.concatDocument.onDidChange(
+      this.onDidChange,
+      this
+    );
+    this.updateCellTracking();
+  }
+
+  notebook: NotebookDocument | undefined = undefined;
+
+  public dispose(): void {
+    this.onDidChangeSubscription.dispose();
+  }
+
+  public isCellOfDocument(uri: Uri): boolean {
+    return this.concatDocument.contains(uri);
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  public save(): Thenable<boolean> {
+    // Not used
+    throw new Error("Not implemented");
+  }
+
+  public lineAt(posOrNumber: Position | number): TextLine {
+    return this.concatDocument.lineAt(posOrNumber);
+  }
+
+  public offsetAt(position: Position): number {
+    return this.concatDocument.offsetAt(position);
+  }
+
+  public positionAt(offset: number): Position {
+    return this.concatDocument.positionAt(offset);
+  }
+
+  public getText(range?: Range | undefined): string {
+    return range
+      ? this.concatDocument.getText(range)
+      : this.concatDocument.getText();
+  }
+
+  public getWordRangeAtPosition(
+    position: Position,
+    regexp?: RegExp | undefined
+  ): Range | undefined {
+    return this.concatDocument.getWordRangeAtPosition(position, regexp);
+  }
+
+  public validateRange(range: Range): Range {
+    return this.concatDocument.validateRange(range);
+  }
+
+  public validatePosition(pos: Position): Position {
+    return this.concatDocument.validatePosition(pos);
+  }
+
+  public locationAt(range: Range): Location {
+    return this.concatDocument.locationAt(range);
+  }
+
+  public getTextDocumentAtPosition(
+    position: Position
+  ): TextDocument | undefined {
+    const location = this.concatDocument.locationAt(position);
+    return this.concatDocument
+      .getComposeDocuments()
+      .find((c) => c.uri === location.uri);
+  }
+
+  private get filteredCells(): NotebookCell[] {
```

Comment:
This is where the real fix is. The logic wasn't taking into account that some of the cells don't count.